component/client-handler/src/main/java/org/wso2/carbon/identity/oauth2/token/handler/clientauth/jwt/PrivateKeyJWTClientAuthenticator.java

```diff
@@ -0,0 +1,178 @@
+        return isValidJWTClientAssertionRequest(oauthJWTAssertionType, oauthJWTAssertion);
+    }
+
+    /**
+     * Retrievs the client ID which is extracted from the JWT.
+     *
+     * @param httpServletRequest
+     * @param bodyParameters
+     * @param oAuthClientAuthnContext
+     * @return jwt 'sub' value as the client id
+     * @throws OAuthClientAuthnException
+     */
+    @Override
+    public String getClientId(HttpServletRequest httpServletRequest, Map<String, List> bodyParameters,
+                              OAuthClientAuthnContext oAuthClientAuthnContext) throws OAuthClientAuthnException {
+
+        String oauthJWTAssertion = getBodyParameters(bodyParameters).get(OAUTH_JWT_ASSERTION);
+        SignedJWT signedJWT = getSignedJWT(oauthJWTAssertion);
+        oAuthClientAuthnContext.addParameter(SIGNED_JWT, signedJWT);
+        ReadOnlyJWTClaimsSet claimsSet = jwtValidator.getClaimSet(signedJWT);
+        return jwtValidator.resolveSubject(claimsSet);
+    }
+
+    private SignedJWT getSignedJWT(String assertion) throws OAuthClientAuthnException {
+
+        SignedJWT signedJWT;
+        if (isEmpty(assertion)) {
+            String errorMessage = "No Valid Assertion was found for " + Constants.OAUTH_JWT_BEARER_GRANT_TYPE;
+            throw new OAuthClientAuthnException(errorMessage, OAuth2ErrorCodes.INVALID_REQUEST);
+        }
+        try {
+            signedJWT = SignedJWT.parse(assertion);
+            logJWT(signedJWT);
+        } catch (ParseException e) {
+            String errorMessage = "Error while parsing the JWT.";
+            throw new OAuthClientAuthnException(errorMessage, OAuth2ErrorCodes.INVALID_REQUEST);
+        }
+        if (signedJWT == null) {
+            String errorMessage = "No Valid Assertion was found for " + OAUTH_JWT_BEARER_GRANT_TYPE;
+            throw new OAuthClientAuthnException(errorMessage, OAuth2ErrorCodes.INVALID_REQUEST);
+        }
+        return signedJWT;
+    }
+
+    private void logJWT(SignedJWT signedJWT) {
+
+        if (log.isDebugEnabled() && IdentityUtil.isTokenLoggable(IdentityConstants.IdentityTokens.ACCESS_TOKEN)) {
+            log.debug(signedJWT);
+        }
+    }
+
+    private boolean isValidJWTClientAssertionRequest(String clientAssertionType, String clientAssertion) {
+
+        if (log.isDebugEnabled()) {
+            log.debug("Authenticate Requested with : " + clientAssertionType + ".");
+            if (IdentityUtil.isTokenLoggable(IdentityConstants.IdentityTokens.ACCESS_TOKEN)) {
+                log.debug(clientAssertion);
+            }
+        }
+        return OAUTH_JWT_BEARER_GRANT_TYPE.equals(clientAssertionType) && isNotEmpty(clientAssertion);
+    }
+
+    private JWTValidator createJWTValidator(String tokenEPAlias, String issuer, boolean preventTokenReuse,
+                                            int rejectBefore) {
+
+        List<Object> mandatoryClaims = new ArrayList<>();
```

Comment:
When is mandatoryClaims populated? 
List<Object> can be changed to List<String>